omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_postgres_validator.rb

```
@@ -0,0 +1,322 @@
+    fail_with err_missing_superuser_password unless cs_pg_attr.has_key? 'db_superuser_password'
+  end
+
+  # We do not support changing from managed to external DB or vice-versa, so the
+  # 'external' flag may not be changed in any scenario - it must be set from the first run
+  # of chef-server-ctl reconfigure.
+  def verify_unchanged_external_flag
+
+    # Given that someone may  move a back-end to a new instance and update the
+    # front-ends with the shared chef-server.rb, we may reasonably expect to
+    # see this value change on the front end - so let's not call that out as an error.
+
+    # This test isn't valid until we're bootstrapped (which could
+    # encompass multiple runs in case of error). Before bootstrap,
+    # the default value of 'false' will be in place, and having a
+    # differing value is valid.
+    return unless OmnibusHelper.has_been_bootstrapped?
+    return unless previous_run
+    return unless backend?
+    puts "EXTERNAL #{cs_pg_attr['external']} vs #{previous_run['postgresql']['external']}"
+    if cs_pg_attr.has_key? 'external' && (cs_pg_attr['external'] != previous_run['postgresql']['external'])
+      fail_with err_cannot_change_external_flag
+    end
+  end
+
+  def common_validation
+    # all nodes are expected to be able to reach the database node
+    # and connect to it - let's make a connection intended to fail
+    # just to verify connectivity to the service.
+    # TODO - might be worth it add post-bootstrap configure tests that
+    # verify health of application database connections?
+    begin
+      connect_as(:invalid_user)
+    rescue ::PG::ConnectionBad => e
+      # A bit messy but PG gem does not expose error codes to us:
+      case e.message
+      when /.*Connection refused.*/
+        fail_with err_postgres_not_available
+      when /.*password authentication failed.*/
+        # This is what we want to see.
+      when /.*no pg_hba.conf entry.*/
```

Unfortunately no. The pg gem gives us two classes of errors, and hides the error codes completely.